src/test/java/frc/lib/swerve/TestSwerveSetpointGenerator.java

```diff
@@ -0,0 +1,137 @@
+    }
+
+    public SwerveSetpoint driveToGoal(SwerveSetpoint prevSetpoint, ChassisSpeeds goal, SwerveSetpointGenerator generator,
+                                      boolean checkAcceleration) {
+        System.out.println("Driving to goal state " + goal);
+        System.out.println("Initial state: " + prevSetpoint);
+        while (!Epsilon.epsilonEquals(CustomTwist2d.fromChassisSpeeds(prevSetpoint.chassisSpeeds), CustomTwist2d.fromChassisSpeeds(goal), Epsilon.EPSILON)) {
+            var newsetpoint = generator.generateSetpoint(KINEMATIC_LIMITS, prevSetpoint, goal, kDt);
+            System.out.println(newsetpoint);
+            SatisfiesConstraints(prevSetpoint, newsetpoint, checkAcceleration);
+            prevSetpoint = newsetpoint;
+        }
+        return prevSetpoint;
+    }
+
+    @Test
+    public void testAccelerationLimits() {
+        SwerveModuleState[] initialStates = {
+                new SwerveModuleState(),
+                new SwerveModuleState(),
+                new SwerveModuleState(),
+                new SwerveModuleState()
+        };
+        SwerveSetpoint setpoint = new SwerveSetpoint(new ChassisSpeeds(), initialStates);
+
+        var generator = new SwerveSetpointGenerator(kKinematics);
+
+        // Just drive straight
+        var goalSpeeds = new ChassisSpeeds(5.0, 0.0, 0.0);
+        setpoint = driveToGoal(setpoint, goalSpeeds, generator, true);
+    }
+
+    @Test
+    public void testGenerateSetpoint() {
+        SwerveModuleState[] initialStates = {
+            new SwerveModuleState(),
+            new SwerveModuleState(),
+            new SwerveModuleState(),
+            new SwerveModuleState()
+        };
+        SwerveSetpoint setpoint = new SwerveSetpoint(new ChassisSpeeds(), initialStates);
+
+        var generator = new SwerveSetpointGenerator(kKinematics);
+
+        var goalSpeeds = new ChassisSpeeds(0.0, 0.0, 1.0);
+        setpoint = driveToGoal(setpoint, goalSpeeds, generator);
+
+        goalSpeeds = new ChassisSpeeds(0.0, 0.0, -1.0);
+        setpoint = driveToGoal(setpoint, goalSpeeds, generator);
+
+        goalSpeeds = new ChassisSpeeds(0.0, 0.0, 0.0);
+        setpoint = driveToGoal(setpoint, goalSpeeds, generator);
+
+        goalSpeeds = new ChassisSpeeds(1.0, 0.0, 0.0);
+        setpoint = driveToGoal(setpoint, goalSpeeds, generator);
+
+        goalSpeeds = new ChassisSpeeds(0.0, 1.0, 0.0);
+        setpoint = driveToGoal(setpoint, goalSpeeds, generator);
+
+        goalSpeeds = new ChassisSpeeds(0.1, -1.0, 0.0);
+        setpoint = driveToGoal(setpoint, goalSpeeds, generator);
+
+        goalSpeeds = new ChassisSpeeds(1.0, -0.5, 0.0);
+        setpoint = driveToGoal(setpoint, goalSpeeds, generator);
+
+        goalSpeeds = new ChassisSpeeds(1.0, 0.4, 0.0);
+        setpoint = driveToGoal(setpoint, goalSpeeds, generator);
+
+        goalSpeeds = new ChassisSpeeds(1.0, 0.0, 1.0);
+        setpoint = driveToGoal(setpoint, goalSpeeds, generator);
+    }

```

Comment:
omg real junits i never heard you talk about these does it work

Reply:
Yes it does work, we can also make a junit of new Robot() I think or at least new robot container.